hibernate-core/src/main/java/org/hibernate/type/CustomType.java

```diff
@@ -216,6 +218,32 @@ public Object disassemble(Object value, SharedSessionContractImplementor session
 		return disassembled;
 	}
 
+	@Override
+	public void addToCacheKey(MutableCacheKeyBuilder cacheKey, Object value, SharedSessionContractImplementor session) {
+		final Serializable disassembled = getUserType().disassemble( (J) value );
+		// Since UserType#disassemble is an optional operation,
+		// we have to handle the fact that it could produce a null value,
+		// in which case we will try to use a converter for disassembling,
+		// or if that doesn't exist, simply use the domain value as is
+		if ( disassembled == null && value != null ) {
+			final BasicValueConverter<J, Object> valueConverter = getUserType().getValueConverter();
+			if ( valueConverter == null ) {
+				cacheKey.addValue( value );
+				cacheKey.addHashCode( Objects.hashCode( value ) );
+			}
+			else {
+				Serializable disassemble = valueConverter.getRelationalJavaType().getMutabilityPlan().disassemble(
+						valueConverter.toRelationalValue( (J) value ),
+						session
+				);
+				cacheKey.addValue( disassemble );
+				cacheKey.addHashCode( valueConverter.getRelationalJavaType().extractHashCode( disassemble ) );
+			}
+		}
+		cacheKey.addValue( disassembled );
+		cacheKey.addHashCode( Objects.hashCode( disassembled ) );
```

Comment:
```suggestion
		final Serializable disassembled = getUserType().disassemble( (J) value );
		// Since UserType#disassemble is an optional operation,
		// we have to handle the fact that it could produce a null value,
		// in which case we will try to use a converter for disassembling,
		// or if that doesn't exist, simply use the domain value as is
		if ( disassembled == null && value != null ) {
			final BasicValueConverter<J, Object> valueConverter = getUserType().getValueConverter();
			if ( valueConverter == null ) {
				cacheKey.addValue( value );
			}
			else {
				Serializable disassemble = valueConverter.getRelationalJavaType().getMutabilityPlan().disassemble(
						valueConverter.toRelationalValue( (J) value ),
						session
				);
				cacheKey.addValue( disassemble );
			}
		}
		else {
			cacheKey.addValue( disassembled );
		}
		cacheKey.addHashCode( getUserType().hashCode( (J) value) );
```